### PR TITLE
implements try catch

### LINKED ISSUE #424: Implement `throw` properly and actually catching errors properly
http://www.byond.com/docs/ref/#/proc/throw

Basic try/catch support got added in https://github.com/wixoaGit/OpenDream/issues/227

But we're still missing `throw` and also it'll never actually *catch* exceptions at the moment.

### PR DESCRIPTION
**Pull request by PaulRitter**
fixes #424

### CODE REVIEW

File: OpenDreamRuntime/Procs/AsyncNativeProc.cs [hunk 1]
```diff
@@ -102,7 +102,8 @@ public sealed class State : ProcState {
                 Result = await _taskFunc(this);
             }
 
-            protected override ProcStatus InternalResume() {
+            public override ProcStatus Resume()
+            {
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/AsyncNativeProc.cs:106]:
```suggestion
            public override ProcStatus Resume() {
```

File: OpenDreamRuntime/DreamThread.cs [hunk 2]
```diff
@@ -203,7 +212,41 @@ public sealed class DreamThread {
...
+                    } catch (DMRuntime dmRuntime) {
+                        if (TryCatchException(dmRuntime.Value)) continue;
+                        HandleException(dmRuntime);
+                        status = ProcStatus.Returned;
+                    } catch (DMError dmError) {
+                        CancelAll();
+                        HandleException(dmError);
+                        status = ProcStatus.Cancelled;
+                    } catch (Exception exception) {
+                        if (TryCatchException(new DreamValue(exception.Message))) continue;
```
Comment (wixoaGit) [inline OpenDreamRuntime/DreamThread.cs:244]: BYOND appears to turn any exception that isn't from a `throw` into an [/exception object](https://www.byond.com/docs/ref/#/exception).

File: OpenDreamRuntime/Procs/DMProc.cs [hunk 3]
```diff
@@ -189,6 +192,9 @@ sealed class DMProcState : ProcState {
         public int ArgumentCount;
         public string? CurrentSource;
         public int CurrentLine;
+        public int CatchPosition = NoTryCatch;
+        public int CatchVarIndex = NoTryCatch;
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/DMProc.cs:196]: These should be reset in either `Dispose()` or `Initialize()`

Proc states can be saved for later reuse, so it's good to not keep values between uses.

File: OpenDreamRuntime/Procs/DMProc.cs [hunk 4]
```diff
@@ -331,6 +348,30 @@ sealed class DMProcState : ProcState {
             return thread;
         }
 
+        public void StartTryBlock(int catchPosition, int catchVarIndex = NoTryCatch) {
+            CatchPosition = catchPosition;
+            CatchVarIndex = catchVarIndex;
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/DMProc.cs:353]: This wouldn't support nested try/catches

File: OpenDreamRuntime/Procs/NativeProc.cs [hunk 5]
```diff
@@ -49,7 +49,8 @@ public sealed class State : ProcState {
                 Arguments = arguments;
             }
 
-            protected override ProcStatus InternalResume() {
+            public override ProcStatus Resume()
+            {
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/NativeProc.cs:53]:
```suggestion
            public override ProcStatus Resume() {
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Comment by wixoaGit** [comment]
The `Try` and `TryNoValue` opcodes will need added to the `ProcDecoder` class, which is used by both the debugger and the decompiler.

**Comment by PaulRitter** [comment]
> The `Try` and `TryNoValue` opcodes will need added to the `ProcDecoder` class, which is used by both the debugger and the decompiler.

Im unsure where i need to add these opcodes in the file, as well as how to test it. Can you ping me on discord with instructions?

File: OpenDreamRuntime/DreamThread.cs [hunk 6]
```diff
@@ -86,16 +86,32 @@ public abstract class DreamProc {
...
+        }
+
+        public DMRuntime(DreamValue value) : base($"'throw' thrown ({value})") {
+            Value = value;
+        }
+    }
+
+    sealed class DMCrashRuntime : DMRuntime {
+
+        public DMCrashRuntime(string message) : base(new DreamValue(message)) { }
```
Comment (wixoaGit) [inline OpenDreamRuntime/DreamThread.cs:104]: Using this base constructor will end up putting `'throw' thrown` at the beginning of the exception message. Probably meant to use the other one?
```suggestion
        public DMCrashRuntime(string message) : base(message) { }
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**